### ISSUE TITLE
Problems with file uploads on remote hosts

### ISSUE THREAD

**Issue by bretrzaun** [minkphp/MinkSelenium2Driver#187]
I can't get remote file uploads to work using Mink with the Selenium2 driver and SauceLabs.
After some research I found the sources below saying one needs to use the "fileDetector".
I am not sure if Mink does already support this. I could not find anything mentioned in the docs and source.

Sources:
- https://saucelabs.com/resources/selenium-file-upload
- http://sauceio.wordpress.com/2014/01/08/remote-file-uploads-with-selenium-php/

**Comment by aik099** [comment]
@robocoder , ping.

**Comment by tompedals** [comment]
I resolved this issue with the following code:

```
$localFile = '/path/to/my/fixture.pdf';
$tempZip = tempnam('', 'WebDriverZip');
$zip = new \ZipArchive();
$zip->open($tempZip, \ZipArchive::CREATE);
$zip->addFile($localFile, basename($localFile));
$zip->close();

$remotePath = $this->getSession()->getDriver()->getWebDriverSession()->file([
  'file' => base64_encode(file_get_contents($tempZip))
]);

$this->attachFileToField('file_input', $remotePath);

unlink($tempZip);
```

I'll try and tidy this up to make it more useful and submit a PR.

**Comment by aik099** [comment]
So the solution is to send actual file in archive to get it uploaded? What file the web server, which selenium goes to will receive, the archive or original file?

**Comment by tompedals** [comment]
@aik099 The `$localFile` is what is attached to the field. Selenium seems to want it sent as a zip file.

**Comment by tompedals** [comment]
I found this solution in the facebook/php-webdriver project.

https://github.com/facebook/php-webdriver/blob/7cfa09f0a50b5e93087a9dc0eb669bc559eb9961/lib/Remote/RemoteWebElement.php#L329

**Comment by aik099** [comment]
Interesting. Could you please send a PR for that? I guess we need a way to determine which kind of upload we want to be used automatically as well.

**Comment by tompedals** [comment]
The way it seems to be done elsewhere is by checking whether the specified file exists on the local filesystem. If it does then it will upload the file to Selenium and use the remote path. If it doesn't exist we assume it is a file on the Selenium server already so just send the keys.

https://github.com/facebook/php-webdriver/blob/7cfa09f0a50b5e93087a9dc0eb669bc559eb9961/lib/Remote/RemoteWebElement.php#L302

If I'm understanding this correctly it means even if you're running a Selenium server locally it will still upload the file and use the remote path. Probably a slight overhead.

Potentially we could check if the `$wdHost` is localhost somehow and use that to determine whether to upload the file or just send keys.

Any opinions on this?

**Comment by aik099** [comment]
> If I'm understanding this correctly it means even if you're running a Selenium server locally it will still upload the file and use the remote path. Probably a slight overhead.

It won't. We have 2 usecases:

1. Mink & Selenium are running on same machine - path to file specified for upload to Mink would exist on selenium side as well
2. Mink & Selenium are running on different machines - path won't exist and it would zip file and transfer it to Selenium machine and use path from there

So it should be fine. In any case tests needs to be created to test both cases. I have no idea how to simulate different machine use case knowing that on Travis tests are running on same machine.

**Comment by tompedals** [comment]
What I mean is in all of the examples I've seen of the file detector component they check if the file exists locally where running Mink, rather than checking if it exists on the Selenium server.

1. Mink & Selenium are running on same machine:
    a. File exists on machine but it will still be uploaded to Selenium and remote path used
    b. File does not exist on machine, so won't upload, but won't exist on Selenium machine so will fail

2. Mink & Selenium are running on different machines:
    a. File exists on Mink machine so will be uploaded to Selenium and remote path used
    b. File does not exist on Mink machine, so won't upload and will just send keys for the path

So most existing use cases where the file exists locally on the Mink machine would now upload the file to Selenium and use the path returned to attach the file to the field.

That is how I understand all of the examples I've seen of this feature so far.

An alternative I had in mind was to determine whether Selenium is running on the same server as Mink and then not upload the file as it's not needed.

**Comment by aik099** [comment]
1a - this can't be. As you can see from https://github.com/minkphp/MinkSelenium2Driver/issues/187#issuecomment-61309499 comment if file exists locally on machine where Mink runs it won't be uploaded as remote one. Anyway you can create PR with changes and tests that will prove either behavior.

**Comment by johnennewdeeson** [comment]
Just wanted to say I was having this problem and the code provided above from 15 Jun works perfectly.

**Comment by aik099** [comment]
@johnennewdeeson , you mean https://github.com/minkphp/MinkSelenium2Driver/issues/187#issuecomment-112106857 comment?

**Comment by aik099** [comment]
Time for PR then. Main point is to automatically determine which kind of upload (local or remote) is needed and execute appropriate code.

**Comment by pfrenssen** [comment]
I've started researching this. The workaround from @tompedals above uses a "file" command that is not officially supported by the webdriver protocol, but it is OK to use this for Selenium only. Selenium themselves use it in their own drivers, see here for example in the Python implementation: https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/remote/webelement.py#L533

It seems indeed that both `php-webdriver` and the official python implementation are checking the filename to see if it exists locally. If it does, then it transfers the file and replaces the filename with the remote path. I think actually the reason for this is because `sendKeys` simulates both the user typing in a normal text input field, or it can also be used to upload a file to a file field. In case this is "normal" keyboard input to a text field then the input wouldn't exist as a file. This seems to indicate a bug though, it looks like if this method is called because you are trying to type a system path in a text field, then it would unnecessarily upload the file.

Ref implementation of `sendKeys()` in Python: https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/remote/webelement.py#L317

**Comment by aik099** [comment]
Fixed in #252.
